### PR TITLE
Add base and startcol to ctx

### LINKED ISSUE #4: startcol should be part of completor's context
core should compute `startcol` and pass it to completor function as part of context

### PR DESCRIPTION
**Pull request by DonnieWest**
Fixes #4 

Probably super clumsy and needs to be fixed, but, hopefully demonstrates the idea

I'm willing to clean this up or integrate any feedback :)

### CODE REVIEW

**Comment by prabirshrestha** [comment]
Out of curiosity for what source do you need this?

**Comment by DonnieWest** [comment]
I'm trying to create a sort of shim between this project and nvim-completion-manager to allow them to share sources. I've a half-working implementation I've been messing with here https://gist.github.com/DonnieWest/505a1ff5edc84345755fb858ac5d93dd

Apologies for the _terrible_ code. I was messing with this late last night and, well, Vimscript is definitely not my best language.

I also understand if this isn't something you'd like to do. This is really just me messing around and, honestly, the best way to solve this is to just patch the upstream projects that I'm trying to support

File: autoload/asyncomplete.vim [hunk 1]
```diff
@@ -235,8 +235,18 @@ function! s:notify_sources_to_refresh(ctx, force) abort
         endif
         if l:refresh
             try
-                call asyncomplete#log('core.s:notify_sources_to_refresh', 'completor()', l:source_name, a:ctx)
-                call s:sources[l:source_name].completor(s:sources[l:source_name], a:ctx)
+                let l:ctx = a:ctx
+                let l:typed = l:ctx['typed']
+                let l:matchpos = s:get_matchpos(s:sources[l:source_name], a:ctx)
```
Comment (prabirshrestha) [inline autoload/asyncomplete.vim]: we are recomputing some of these, can you only compute if it hasn't been computed yet? You can most likely use `if !exits('l:matchpos')` to check if has been completed or not.

File: autoload/asyncomplete.vim [hunk 2]
```diff
@@ -235,8 +235,18 @@ function! s:notify_sources_to_refresh(ctx, force) abort
         endif
         if l:refresh
             try
-                call asyncomplete#log('core.s:notify_sources_to_refresh', 'completor()', l:source_name, a:ctx)
-                call s:sources[l:source_name].completor(s:sources[l:source_name], a:ctx)
+                let l:ctx = a:ctx
```
Comment (prabirshrestha) [inline autoload/asyncomplete.vim]: need to make a copy if we are modifying it since it is in loop. `let l:ctx = copy(a:copy)`

**Comment by prabirshrestha** [comment]
I have actually always wanted `startcol` to be computed by asyncomplete so this is a good feature.

**Comment by DonnieWest** [comment]
I think I've incorporated your feedback

**Comment by prabirshrestha** [comment]
@DonnieWest I have merged it thanks for the PR. Please do keep me informed about your experiment on shims  and best of luck 😄

**Comment by DonnieWest** [comment]
Hrm, after further review here I think I'm calculating the startcol incorrectly?

My updated shim now calculates based off of a word regex here https://gist.github.com/DonnieWest/505a1ff5edc84345755fb858ac5d93dd#file-cm-vim-L40 and sources work as expected

I'm noticing that nvim completion manager uses 'word_pattern' to calculate this, with default values found here https://github.com/roxma/nvim-completion-manager/blob/e724a442072261993ca503e969d2cb25722ab1d2/pythonx/cm_default.py but I'm not sure the best way to accomplish the same thing in Vimscript

What do you think would be the best way to fix this @prabirshrestha ?

In other news though, my shim works quite well against the master branch. There are some other bugs against the TextChangedP branch but I'll submit those as issues shortly

**Comment by prabirshrestha** [comment]
s:get_matchpos() Uses the refresh_pattern similar to ncm so you should be able to calculate the startcol.

**Comment by DonnieWest** [comment]
Aha. Yeah, that would've been better of me. Apologies, I'll try to submit a new PR when I get some time

**Comment by DonnieWest** [comment]
I take that back - ncm uses two separate patterns to manage its completions:

1. `word_pattern`
2. `refresh_pattern`

The current code is using `refresh_pattern`, but could be cleaned up to make that clearer. Alternatively, we could integrate `word_pattern` into asyncomplete to allow sources to override that

What would you prefer @prabirshrestha ?

**Comment by prabirshrestha** [comment]
@DonnieWest Would it be possible to add the startcol on your shim instead for now?
I'm thinking of reverting this PR for now so that I can ship the v2 first while we think about this.

Thoughts?

**Comment by DonnieWest** [comment]
I'm already doing that, so, feel free to revert

**Comment by prabirshrestha** [comment]
@DonnieWest reverted it. we can tackle it again once we merge it to master.

**Pull request merged**